src/common/hist_util.h

```diff
@@ -190,6 +190,101 @@ size_t DeviceSketch(int device,
                     DMatrix* dmat,
                     HistogramCuts* hmat);
 
+
+enum BinBounds {
+  UINT8_BINS_TYPE = 0,
+  UINT16_BINS_TYPE,
+  UINT32_BINS_TYPE,
+};
+
+struct Index {
+  Index(): binBound_(UINT8_BINS_TYPE), p_(1), disp_ptr_(nullptr) {
+    setBinBound(binBound_);
+  }
+  Index(const Index& i) = delete;
+  Index& operator=(Index i) = delete;
+  Index(Index&& i) = delete;
+  Index& operator=(Index&& i) = delete;
+  uint32_t operator[](size_t i) const {
+    if (disp_ptr_ != nullptr) {
+      return func_(data_ptr_, i) + disp_ptr_[i%p_];
+    } else {
+      return func_(data_ptr_, i);
+    }
+  }
+  void setBinBound(BinBounds binBound) {
+    binBound_ = binBound;
+    switch (binBound) {
+      case UINT8_BINS_TYPE:
+        func_ = &getValueFromUint8;
+        break;
+      case UINT16_BINS_TYPE:
+        func_ = &getValueFromUint16;
+        break;
+      case UINT32_BINS_TYPE:
+        func_ = &getValueFromUint32;
+        break;
+      default:
+        CHECK(binBound == UINT8_BINS_TYPE  ||
+              binBound == UINT16_BINS_TYPE ||
+              binBound == UINT32_BINS_TYPE);
+    }
+  }
+  BinBounds getBinBound() const {
+    return binBound_;
+  }
+  template<typename T>
+  T* data() const {
+    return static_cast<T*>(data_ptr_);
+  }
+  uint32_t* disp() const {
+    return disp_ptr_;
+  }
+  size_t dispSize() const {
+    return disp_.size();
+  }
+  size_t size() const {
+    return data_.size() / (1 << binBound_);
+  }
+  void resize(const size_t nBytesData) {
+    data_.resize(nBytesData);
+    data_ptr_ = reinterpret_cast<void*>(data_.data());
+  }
+  void resizeDisp(const size_t nDisps) {
+    disp_.resize(nDisps);
+    disp_ptr_ = disp_.data();
+    p_ = nDisps;
+  }
+  std::vector<uint8_t>::const_iterator begin() const {
+    return data_.begin();
+  }
+  std::vector<uint8_t>::const_iterator end() const {
+    return data_.end();
+  }
+
+ private:
+  static uint32_t getValueFromUint8(void *t, size_t i) {
+    return reinterpret_cast<uint8_t*>(t)[i];
+  }
+  static uint32_t getValueFromUint16(void* t, size_t i) {
+    return reinterpret_cast<uint16_t*>(t)[i];
+  }
+  static uint32_t getValueFromUint32(void* t, size_t i) {
+    return reinterpret_cast<uint32_t*>(t)[i];
+  }
+
+  typedef uint32_t (*Func)(void*, size_t);
+
+  std::vector<uint8_t> data_;
+  std::vector<uint32_t> disp_;
```

Comment:
Please change it to a better name.

Reply:
Agree, don't know what disp means after a quick read.

Reply:
Fixed. Thanks for valuable comment.